### PR TITLE
Stratify enumerative instantiation

### PR DESCRIPTION
**Pull request by ajreynol**
This adds an option (--fs-stratify) to stratify the effort levels (relevant domain, arbitrary ground term) in enumerative instantiation globally instead of locally per quantified formula.  The upside of this option is that arbitrary ground term can be very expensive, which this option avoids. The downside of this option is that it may be "unfair": quantified formulas with no tuples of terms in their relevant domains may never be instantiated.  It is disabled by default.

Indentation in the code changed so the changes appear larger than they are.

### CODE REVIEW

**Comment by HanielB** [comment]
Can you post a summary of your evaluation for future reference?

File: src/theory/quantifiers/inst_strategy_enumerative.cpp [hunk 1]
```diff
@@ -70,225 +70,244 @@ void InstStrategyEnum::check(Theory::Effort e, QEffort quant_e)
...
-    int addedLemmas = 0;
-    for (unsigned i = 0;
-         i < d_quantEngine->getModel()->getNumAssertedQuantifiers();
-         i++)
+    return;
+  }
+  double clSet = 0;
+  if (Trace.isOn("fs-engine"))
+  {
+    clSet = double(clock()) / double(CLOCKS_PER_SEC);
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_enumerative.cpp]: Don't we want to be using `TimerStat` for this kind of stuff?

File: src/theory/quantifiers/inst_strategy_enumerative.cpp [hunk 2]
```diff
@@ -70,225 +70,244 @@ void InstStrategyEnum::check(Theory::Effort e, QEffort quant_e)
...
-  unsigned rend = fullEffort ? 1 : rstart;
-  for (unsigned r = rstart; r <= rend; r++)
+  unsigned final_max_i = 0;
+  std::vector<unsigned> maxs;
+  std::vector<bool> max_zero;
+  bool has_zero = false;
+  std::map<TypeNode, std::vector<Node> > term_db_list;
+  std::vector<TypeNode> ftypes;
+  // iterate over substitutions for variables
+  for (unsigned i = 0; i < f[0].getNumChildren(); i++)
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_enumerative.cpp]: cache size

File: src/theory/quantifiers/inst_strategy_enumerative.cpp [hunk 3]
```diff
@@ -70,225 +70,244 @@ void InstStrategyEnum::check(Theory::Effort e, QEffort quant_e)
...
+    ftypes.push_back(tn);
+    unsigned ts;
+    if (isRd)
     {
-      if (r == 0)
+      ts = rd->getRDomain(f, i)->d_terms.size();
+    }
+    else
+    {
+      ts = d_quantEngine->getTermDatabase()->getNumTypeGroundTerms(tn);
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_enumerative.cpp]: retrieve termdb only once outside loop

File: src/theory/quantifiers/inst_strategy_enumerative.cpp [hunk 4]
```diff
@@ -70,225 +70,244 @@ void InstStrategyEnum::check(Theory::Effort e, QEffort quant_e)
...
-        rd->compute();
-        Trace("inst-alg-debug") << "...finished" << std::endl;
+        std::map<Node, Node> reps_found;
+        for (unsigned j = 0; j < ts; j++)
+        {
+          Node gt =
+              d_quantEngine->getTermDatabase()->getTypeGroundTerm(ftypes[i], j);
+          if (!options::cbqi() || !quantifiers::TermUtil::hasInstConstAttr(gt))
+          {
+            Node rep = d_quantEngine->getEqualityQuery()->getRepresentative(gt);
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_enumerative.cpp]: ditto

File: src/theory/quantifiers/inst_strategy_enumerative.cpp [hunk 5]
```diff
@@ -70,225 +70,244 @@ void InstStrategyEnum::check(Theory::Effort e, QEffort quant_e)
...
+    max_zero.push_back(fullEffort && ts == 0);
+    ts = (fullEffort && ts == 0) ? 1 : ts;
+    Trace("inst-alg-rd") << "Variable " << i << " has " << ts
+                         << " in relevant domain." << std::endl;
+    if (ts == 0)
+    {
+      has_zero = true;
+      break;
+    }
+    else
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_enumerative.cpp]: no need for this `else`

File: src/theory/quantifiers/inst_strategy_enumerative.cpp [hunk 6]
```diff
@@ -70,225 +70,244 @@ void InstStrategyEnum::check(Theory::Effort e, QEffort quant_e)
...
-          ts = rd->getRDomain(f, i)->d_terms.size();
-        }
-        else
+        while (index >= 0 && index < (int)f[0].getNumChildren())
         {
-          ts = d_quantEngine->getTermDatabase()->getNumTypeGroundTerms(tn);
-          std::map<TypeNode, std::vector<Node> >::iterator ittd =
-              term_db_list.find(tn);
-          if (ittd == term_db_list.end())
+          if (index == (int)childIndex.size())
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_enumerative.cpp]: static_cast

File: src/theory/quantifiers/inst_strategy_enumerative.cpp [hunk 7]
```diff
@@ -70,225 +70,244 @@ void InstStrategyEnum::check(Theory::Effort e, QEffort quant_e)
...
-        while (max_i <= final_max_i)
-        {
-          Trace("inst-alg-rd") << "Try stage " << max_i << "..." << std::endl;
-          std::vector<unsigned> childIndex;
-          int index = 0;
-          do
+          Trace("inst-alg-rd") << "}" << std::endl;
+          // try instantiation
+          std::vector<Node> terms;
+          for (unsigned i = 0; i < f[0].getNumChildren(); i++)
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_enumerative.cpp]: cache size

File: src/theory/quantifiers/inst_strategy_enumerative.cpp [hunk 8]
```diff
@@ -70,225 +70,244 @@ void InstStrategyEnum::check(Theory::Effort e, QEffort quant_e)
...
-          break;
-        }
-        else
+        success = index >= 0;
+        if (success)
         {
-          maxs.push_back(ts);
-          if (ts > final_max_i)
+          Trace("inst-alg-rd") << "Try instantiation { ";
+          for (unsigned j = 0; j < childIndex.size(); j++)
```
Comment (HanielB) [inline src/theory/quantifiers/inst_strategy_enumerative.cpp]: cache size

**Comment by ajreynol** [comment]
diff TPTP benchmarks:
```

benchmark | without | time | with
-- | -- | -- | --
Problems/ALG/ALG226+1.p | t/o | 30.014 | unsat
Problems/CAT/CAT019-3.p | unsat | 0.0157021 | t/o
Problems/DAT/DAT085=1.p | unsat | 0.0339509 | t/o
Problems/DAT/DAT089=1.p | unsat | 0.0299599 | t/o
Problems/DAT/DAT091=1.p | unsat | 0.027093 | t/o
Problems/DAT/DAT092=1.p | unsat | 0.0251921 | t/o
Problems/DAT/DAT102=1.p | unsat | 0.033195 | t/o
Problems/DAT/DAT332=1.p | unsat | 0.0889039 | t/o
Problems/GRP/GRP462-1.p | unsat | 23.5841 | t/o
Problems/GRP/GRP527-1.p | unsat | 7.94431 | t/o
Problems/HWV/HWV068-1.p | t/o | 30.0384 | unsat
Problems/KRS/KRS257+1.p | unsat | 0.497138 | t/o
Problems/KRS/KRS258+1.p | unsat | 0.512554 | t/o
Problems/KRS/KRS261+1.p | unsat | 0.625129 | t/o
Problems/LAT/LAT026-1.p | unsat | 14.9888 | t/o
Problems/NLP/NLP220+1.p | t/o | 30.0406 | GaveUp
Problems/NUM/NUM858=1.p | unsat | 0.224754 | t/o
Problems/PLA/PLA017-1.p | t/o | 30.0396 | unsat
Problems/PUZ/PUZ001+2.p | unsat | 0.189545 | t/o
Problems/SET/SET169+3.p | unsat | 2.6374 | t/o
Problems/SET/SET171+3.p | unsat | 2.41125 | t/o
Problems/SET/SET609+3.p | unsat | 2.35878 | t/o
Problems/SET/SET613+3.p | unsat | 3.1026 | t/o
Problems/SET/SET614+3.p | unsat | 0.898652 | t/o
Problems/SET/SET615+3.p | unsat | 1.80109 | t/o
Problems/SET/SET634+3.p | unsat | 1.19462 | t/o
Problems/SET/SET889+1.p | unsat | 1.05426 | t/o
Problems/SET/SET919+1.p | unsat | 1.09665 | t/o
Problems/SET/SET935+1.p | unsat | 7.48481 | t/o
Problems/SET/SET938+1.p | unsat | 12.3056 | t/o
Problems/SET/SET949+1.p | unsat | 18.7061 | t/o
Problems/SET/SET957+1.p | unsat | 8.66554 | t/o
Problems/SET/SET959+1.p | unsat | 0.212864 | t/o
Problems/SET/SET986+1.p | unsat | 1.83453 | t/o
Problems/SEU/SEU126+1.p | unsat | 0.186179 | t/o
Problems/SEU/SEU135+1.p | unsat | 5.1395 | t/o
Problems/SEU/SEU136+1.p | unsat | 5.28115 | t/o
Problems/SEU/SEU137+1.p | unsat | 4.42803 | t/o
Problems/SEU/SEU138+1.p | unsat | 3.59317 | t/o
Problems/SEU/SEU141+1.p | unsat | 1.19989 | t/o
Problems/SEU/SEU147+2.p | unsat | 0.913573 | t/o
Problems/SEV/SEV422=1.p | unsat | 0.11196 | t/o
Problems/SEV/SEV424=1.p | unsat | 25.2673 | t/o
Problems/SEV/SEV425=1.p | unsat | 29.493 | t/o
Problems/SWV/SWV255-2.p | unsat | 0.213095 | t/o
Problems/SWV/SWV411+1.p | unsat | 7.7313 | t/o
Problems/SWV/SWV420-1.005.p | t/o | 30.062 | unsat
Problems/SWV/SWV920-1.p | GaveUp | 2.5742 | t/o
Problems/SWW/SWW677=1.p | unsat | 12.3464 | t/o
Problems/SYN/SYN702-1.p | t/o | 30.0143 | unsat
Problems/SYN/SYN833-1.p | t/o | 30.0169 | unsat
Problems/SYN/SYN836-1.p | t/o | 30.0212 | unsat
Problems/SYN/SYN845-1.p | t/o | 30.0336 | unsat
Problems/SYN/SYN857-1.p | t/o | 30.0261 | unsat
Problems/SYN/SYN869-1.p | t/o | 30.2789 | unsat
Problems/SYN/SYN871-1.p | t/o | 30.0537 | unsat
Problems/SYN/SYN873-1.p | t/o | 30.0304 | unsat
Problems/SYN/SYN874-1.p | t/o | 30.0242 | unsat
Problems/SYN/SYN875-1.p | t/o | 30.0249 | unsat
Problems/SYN/SYN876-1.p | t/o | 30.0794 | unsat
Problems/SYN/SYN877-1.p | t/o | 30.0234 | unsat
Problems/SYN/SYN878-1.p | t/o | 30.0225 | unsat
Problems/SYN/SYN879-1.p | t/o | 30.0319 | unsat
Problems/SYN/SYN880-1.p | t/o | 30.1374 | unsat
Problems/SYN/SYN882-1.p | t/o | 30.0281 | unsat
Problems/SYN/SYN883-1.p | t/o | 30.0188 | unsat
Problems/SYN/SYN889-1.p | t/o | 30.0247 | unsat
Problems/SYN/SYN892-1.p | t/o | 30.0195 | unsat
Problems/SYN/SYN894-1.p | t/o | 30.0196 | unsat
Problems/SYN/SYN895-1.p | t/o | 30.0387 | unsat
Problems/SYO/SYO607+1.p | unsat | 0.295793 | t/o
Problems/SYO/SYO630-1.p | t/o | 30.0308 | unsat
Problems/SYO/SYO631-1.p | t/o | 30.0193 | unsat
Problems/TOP/TOP005-2.p | t/o | 30.0185 | unsat

```

**Comment by ajreynol** [comment]
Total +28-44

File: src/theory/quantifiers/inst_strategy_enumerative.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/inst_strategy_enumerative.cpp]: I dont know, but this is outside of the scope of this PR (this is my convention in quantifiers).

**Pull request merged**